### PR TITLE
Fixes for out-of-range Bitmap references

### PR DESCRIPTION
**Pull request by PeteCA**
Nobody seems to get Bitmap dimensions right!  I happened to be working with a power-of-2 sized image, and hit _two_ separate exceptions!  GaussianBlur and Sharpness both crashed (they use the same code), and trying to scale up x 2 did the same.  Both seem to be in Heikke's original code (but don't crash BeOS).  They are both only off-by-one accesses, and don't even trigger a Haiku exception every time.  They were a pain to track down, but I think the fixes are correct.

(In ScaleManipulator.cpp, I also changed "..._bpr" variables to "..._pxpr", because they are  not bytes-per-row, but 32-bit pixel counts.)

### CODE REVIEW

**Comment by humdingerb** [comment]
FWIW, I tried scaling with the un-fixed version a few times and it didn't crash.
Anyway, if @waddlesplash agrees with your changes, who am I to stand in your way? :)

Just one comment in the review...

File: addons/UtilityClasses/ImageProcessingLibrary.cpp [hunk 1]
```diff
@@ -4,7 +4,7 @@
  *
  * Authors:
  * 		Heikki Suhonen <heikki.suhonen@gmail.com>
- *
+ *	Array overflow fix 2017/7/18  Pete G.
```
Comment (humdingerb) [inline addons/UtilityClasses/ImageProcessingLibrary.cpp]: You wanna change that to be roughly in the same format as Haikki's entry? One more indent, name, email (if you want) and "Array overflow fix (18-07.2017)" in brackets, maybe.

**Comment by PeteCA** [comment]
On Tue, Jul 18, 2017 at 11:08:48PM -0700, humdinger wrote:
> FWIW, I tried scaling with the un-fixed version a few times and it didn't crash.
> Anyway, if @waddlesplash agrees with your changes, who am I to stand in your way? :)
> 
Yes, it doesn't crash every time, and it only happens when the width is a power of 3  (I guess something to do with the precise way memory is allocated), but it was consistent enough to be really frustrating!

**Comment by PeteCA** [comment]
On Tue, Jul 18, 2017 at 11:10:29PM -0700, humdinger wrote:
> 
> You wanna change that to be roughly in the same format as Haikki's entry? One more indent, name, email (if you want) and "Array overflow fix (18-07.2017)" in brackets, maybe.
> 
Hmm, no -- I think I'd just rather remove it...  Not really necessary anyway.  I think I would always want to standardize on that date format I used (never seen your variant before!); it's internationally understandable and becoming common.
(I didn't indent it because I don't consider myself an "author" there. It was more an annotation.)

**Comment by humdingerb** [comment]
I typo'ed my date, wanted to write 18.07.2017. But you're more correct anyway, according to ISO 8601 it's YYYY-MM-DD, so 2017-07-18.
Removing is of course the cleanest solution... :)

**Comment by PeteCA** [comment]
I'm now glad that nobody has merged this yet... (:-/)  My fix works, but I've just realized a much cleaner one that I'll replace it with when I get a chance (later today when I'm on my machine).  I'll explain then.

**Comment by PeteCA** [comment]
Mfff... My "cleaner solution" is stymied by floating-point precision!

Probably good to explain the original problem.   When an image is to be scaled up, as it steps through the destination row -- say 512 pixels, therefore 0..511 -- it needs to step through the source -- 256 wide, say -- and pick suitable corresponding pixels.  It computes a suitable float step size by dividing the source width by the destination width, giving in this case a  step of 0.5.  However it doesn't just grab a single pixel; it gets the integer floor and ceiling indexes of the current floating offset, and averages the retrieved pixels.  The problem is that if it steps from 0 to 511, the final offset is 255.5, the ceil of which is 256!  Oops!  Not fatal until the very last row, and the memory allocation must be exact, but the above values crash consistently.

I thought I could do better by dividing the right-hand pixel indexes, rather than the widths (255.0/511 in the above), which works fine in theory, but it turns out the final offset ends up as something like 255.0049, which still has a ceil of 256!  Damn...

So I'll leave the current fix alone, as it seems solid.  There are still 'edge effects' if right or bottom pixels differ greatly from the  adjacent ones (because it's averaging them) but I can't see any consistent way to improve that.  The fix doesn't make anything worse.

I've removed my annotation at the head of ImageProcessingLibrary.cpp

**Comment by PeteCA** [comment]
Sorry, folks.  My scaling fix wasn't good enough!  I was too specific to scaling exactly by 2.  It would still crash if I scaled by a bit less than that (so the step was larger than 0.5).  The check is now strict on the final offset value.  I've tried a lot of possibly dangerous values and can't make it crash now.

**Comment by humdingerb** [comment]
Don't sweat it, Pete. :)
You think it's now good to merge, or should be let it simmer for a few more days? How long is the mean incubation period for bugs...? :)

**Comment by PeteCA** [comment]
Infuriating, though... (:-/)  I'm pretty certain it's correct, now; the logic seems right,  and as I say I gave it a bit of a torture test.  Any future crashes are likely to be from something else.  I'd say merge it, and then I can try to resync my own master.  Thanks.

**Comment by humdingerb** [comment]
Thanks for your perseverance. :)

**Comment by PeteCA** [comment]
I get this horrible itch when things don't work.... (:-))

Thanks for merging!

**Pull request merged**